mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/util/HierarchyScanner.java

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.mapper.processor.util;
+
+import com.datastax.oss.driver.api.mapper.annotations.HierarchyScanStrategy;
+import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableSet;
+import com.datastax.oss.driver.shaded.guava.common.collect.Sets;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.util.Types;
+
+/** Provides mechanisms for building and traversing a class/interfaces hierarchy. */
+public class HierarchyScanner {
+
+  // placeholder class for resolving the default HierarchyScanStrategy
+  @HierarchyScanStrategy
+  private static final class ClassForDefaultScanStrategy {}
+
+  /**
+   * Resolves the type hierarchy for the given type by first looking for a {@link
+   * HierarchyScanStrategy}-annotated class or interface in the given types hierarchy. The hierarchy
+   * is traversed until a {@link HierarchyScanStrategy} is encountered in a depth-first manner as
+   * follows:
+   *
+   * <ol>
+   *   <li>Initialize <code>i</code> as an empty set
+   *   <li>Visit typeElement (<code>e</code>)
+   *   <li>Visit <code>e</code>'s parent class (<code>p</code>), if <code>p</code> is null, stop
+   *       traversing after evaluating remaining interfaces
+   *   <li>Visit <code>e</code>'s interfaces, and record those interfaces' parents for later use (
+   *       <code>ni</code>)
+   *   <li>Visit <code>i</code>, and append those interface's parents to <code>ni</code> for later
+   *       use)
+   *   <li>If <code>p != null</code> </code>Set <code>e := p, i := ni</code> and repeat starting at
+   *       step 3
+   *   <li>Visit <code>ni</code> interfaces and their parents until we've reached root
+   * </ol>
```

Comment:
These explanations would be a bit easier to follow if you named your variables the same as in the code: `e` => `typeElement`, `i` => `interfacesToScan`, etc.